packages/eslint-plugin/src/rules/consistent-indexed-object-style.ts

```diff
@@ -0,0 +1,118 @@
+import { createRule } from '../util';
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+export default createRule({
+  name: 'consistent-indexed-object-style',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Enforce or disallow the use of the record type',
+      category: 'Stylistic Issues',
+      // too opinionated to be recommended
+      recommended: false,
+    },
+    messages: {
+      preferRecord: 'A record is preferred over an index signature',
+      preferIndexSignature: 'An index signature is preferred over a record.',
+    },
+    fixable: 'code',
+    schema: [
+      {
+        enum: ['always', 'never'],
+      },
+    ],
+  },
+  defaultOptions: ['always'],
+  create(context) {
+    const sourceCode = context.getSourceCode();
+
+    if (context.options[0] === 'never') {
+      return {
+        TSTypeReference(node): void {
+          const typeName = node.typeName as TSESTree.Identifier;
+          if (typeName.name !== 'Record') {
+            return;
+          }
+
+          const params = node.typeParameters?.params;
+          if (params?.length !== 2) {
+            return;
+          }
+
+          context.report({
+            node,
+            messageId: 'preferIndexSignature',
+            fix(fixer) {
+              const key = sourceCode.getText(params[0]);
+              const type = sourceCode.getText(params[1]);
+              return fixer.replaceTextRange(
+                node.range,
+                `{ [key: ${key}]: ${type} }`,
+              );
+            },
+          });
+        },
+      };
+    }
+
+    /**
+     * Convert an index signature node to record code as string.
+     */
+    function toRecord(node: TSESTree.TSIndexSignature): string {
+      const parameter = node.parameters[0] as TSESTree.Identifier;
+      const key = sourceCode.getText(parameter.typeAnnotation!.typeAnnotation);
+      const value = sourceCode.getText(node.typeAnnotation!.typeAnnotation);
```

Comment:
these lines will crash on this code which is syntactically valid:
```ts
type T1 = { [k]: any };
type T2 = { [k: string]: any };
type T3 = { [k] };
```

TS will error on this code, but we need to ensure we don't crash on this code.

Reply:
The rule didn’t crash on these rules. Do you want me to add test cases for `T1` and `T3` anyway?

How is `T2` invalid? This looks like the most basic case handled by this rule.

Reply:
Sorry - T2 was supposed to be

```ts
type T2 = { [k: string] };
```

Reply:
I added a test and was actually able to deduplicate even more logic now.